mqtt/mqtt-edgehub/src/command.rs

```diff
@@ -0,0 +1,248 @@
+
+        Ok(CommandHandler {
+            broker_handle,
+            client,
+        })
+    }
+
+    pub fn shutdown_handle(&self) -> Result<ShutdownHandle, ShutdownError> {
+        self.client
+            .shutdown_handle()
+            .map_or(Err(ShutdownError::ClientDoesNotExist), |shutdown_handle| {
+                Ok(ShutdownHandle(shutdown_handle))
+            })
+    }
+
+    pub async fn run(mut self) {
+        self.wait_for_suback().await;
+
+        self.listen_for_disconnect().await;
+    }
+
+    async fn wait_for_suback(&mut self) {
+        while let Some(event_or_err) = self.client.next().await {
+            if let Err(e) = event_or_err {
+                warn!(message = "error reading event from command handler client", error = %e);
+            } else if let Ok(Event::SubscriptionUpdates(subscriptions)) = event_or_err {
+                for acked_subscription in subscriptions {
+                    if let SubscriptionUpdateEvent::Subscribe(sub) = acked_subscription {
+                        if sub.topic_filter == TOPIC_FILTER {
+                            return;
+                        }
+                    }
+                }
+            } else {
+                warn!("received unexpected event when waiting for command handler suback");
+            }
+        }
+    }
+
+    async fn listen_for_disconnect(&mut self) {
+        while let Some(Ok(event)) = self.client.next().await {
+            if let Err(e) = self.handle_event(event).await {
+                warn!(message = "failed to disconnect client", error = %e);
+            }
+        }
+    }
+
+    async fn handle_event(&mut self, event: Event) -> Result<(), DisconnectClientError> {
+        if let Event::Publication(publication) = event {
+            let topic_name = publication.topic_name.as_ref();
+            let client_id = parse_client_id(topic_name)?;
+
+            info!("received disconnection request for client {}", client_id);
+
+            if let Err(e) = self
+                .broker_handle
+                .send(Message::System(SystemEvent::ForceClientDisconnect(
+                    client_id.into(),
+                )))
+                .await
+            {
+                return Err(DisconnectClientError::SignalError(e));
+            }
+
+            info!("succeeded sending broker signal to disconnect client");
+        }
+
+        Ok(())
+    }
+}
+
+fn parse_client_id(topic_name: &str) -> Result<String, DisconnectClientError> {
+    lazy_static! {
+        static ref REGEX: Regex =
+            Regex::new(CLIENT_EXTRACTION_REGEX).expect("failed to create new Regex from pattern");
+    }
+
+    let captures = REGEX
+        .captures(topic_name.as_ref())
+        .ok_or_else(DisconnectClientError::RegexFailure)?;
+
+    let value = captures
+        .get(1)
+        .ok_or_else(DisconnectClientError::RegexFailure)?;
+
+    let client_id = value.as_str();
+    match client_id {
```

Comment:
redundant allocations
```rust 
    if value.empty() {
         Err(DisconnectClientError::NoClientId())
  } else {
        Ok(value)
    }
```

Reply:
```value``` is a match type not a str type, so there is no method empty()